Simplify annotation handling by using the flatten method:

dbartenstein opened the pull request:
https://github.com/django/django/blob/f42ccdd835e5b3f0914b5e6f87621c648136ea36/django/db/models/expressions.py#L370

Handle **annotation** cases:
- when `Subquery` is part of the `When`
- when `Coalesce` is used
- when `RawSQL` is used

[//]: # (Thanks for helping us out! Cache invalidation sucks, so your help is greatly appreciated!)

## Description
Expressions can contain multiple elements and all of them have to be checked.

[//]: # (What're you proposing?)


## Rationale
Ensure that all elements are checked for queries.

[//]: # (Why should we implement it?)

Andrew-Chen-Wang commented:
@dbartenstein seems like flatten is not part of Django 2.2

Andrew-Chen-Wang commented:
@dbartenstein flatten has a if statement checking if there's a method flatten for each node. But that's only in Django 3.2. Do you mind just copying Django's code to here? That would solve the issue.

dbartenstein commented:
> @dbartenstein flatten has a if statement checking if there's a method flatten for each node. But that's only in Django 3.2. Do you mind just copying Django's code to here? That would solve the issue.

@Andrew-Chen-Wang sure! I am just wondering if most of the code in `_get_tables` could be refactored to loop over the flattened nodes instead? What do you think?

Andrew-Chen-Wang commented on cachalot/utils.py:
```diff
@@ -178,18 +179,32 @@ def __update_annotated_subquery(_annotation: Subquery):
+                if isinstance(element, Subquery):
+                    __update_annotated_subquery(element)
+                elif isinstance(element, RawSQL):
+                    sql = repr(element).lower()
```
Why not use `as_sql`? I thought the repr includes the class name: https://github.com/django/django/blob/bf5abf1bdcedb15e949db419c61eeec7c88414ea/django/db/models/expressions.py#L812

dbartenstein replied:
@Andrew-Chen-Wang: yes. Instead of calling `as_sql` with fake arguments (as they are not used in the method) we could also directly access the sql attribute of `RawSQL`. https://github.com/django/django/blob/f42ccdd835e5b3f0914b5e6f87621c648136ea36/django/db/models/expressions.py#L815

What do you reckon?

Andrew-Chen-Wang commented on cachalot/utils.py:
```diff
@@ -178,18 +179,32 @@ def __update_annotated_subquery(_annotation: Subquery):
+                if isinstance(element, Subquery):
+                    __update_annotated_subquery(element)
+                elif isinstance(element, RawSQL):
+                    sql = element.as_sql(None, None)[0].lower()
```
I'd assume the compiler with the db engine is already attached, but just in case is there a method to check for db alias like in L233?

```sql = query.get_compiler(db_alias).as_sql()[0].lower()```

Andrew-Chen-Wang replied:
Nvm I see the params are actually the compiler...

Andrew-Chen-Wang reviewed (approved):
LGTM! Thanks a ton!

Andrew-Chen-Wang commented:
@dbartenstein I changed the `flatten` function since you were using `element` instead of the given `expression` arg. Can you verify that the behavior is still correct? I'm not entirely sure how the behavior was correct before but... 🤷

dbartenstein commented:
> @dbartenstein I changed the `flatten` function since you were using `element` instead of the given `expression` arg. Can you verify that the behavior is still correct? I'm not entirely sure how the behavior was correct before but... shrug

@Andrew-Chen-Wang that looks good! I also changed other occurrences of `element` to `expression`: it’s all about cache invalidation and naming ;-)

Andrew-Chen-Wang commented:
Thanks a ton for this!

dbartenstein commented:
> Thanks a ton for this!

You’re welcome!
